includes/Story_Renderer/HTML.php

```diff
@@ -111,82 +72,45 @@ public function __construct( Story $story ) {
 	 * @return string The complete HTML markup for the story.
 	 */
 	public function render() {
-		$markup = self::DEFAULT_DOCTYPE . $this->story->get_markup();
-		$markup = $this->adapt_encoding( $markup );
+		$markup = $this->story->get_markup();
 		$markup = $this->replace_html_head( $markup );
-		$markup = $this->remove_noscript_amp_boilerplate( $markup );
 
-		$this->document = $this->load_html( $markup );
+		// TODO: What if transformation failed?
+		$document = Document::fromHtml( $markup, get_bloginfo( 'charset' ) );
+
+		// This  should never actually happen.
+		if ( ! $document ) {
+			wp_die(
+				esc_html__( 'There was an error generating the web story, probably because of a server misconfiguration. Try contacting your hosting provider or open a new support request.', 'web-stories' ),
+				esc_html__( 'Web Stories', 'web-stories' ),
+				[
+					'response'  => 500,
+					'link_url'  => esc_url( __( 'https://wordpress.org/support/plugin/web-stories/', 'web-stories' ) ),
+					'link_text' => esc_html__( 'Visit Support Forums', 'web-stories' ),
+				]
+			);
+		}
+
+		$this->document = $document;
 
-		// Run all further transformations on the DOMDocument.
+		// Run all further transformations on the Document instance.
+		// These need to run before sanitization so it works even when
+		// sanitization is done by the AMP plugin.
 
-		$this->add_noscript_amp_boilerplate();
 		$this->transform_html_start_tag();
 		$this->transform_a_tags();
 		$this->insert_analytics_configuration();
```

Comment:
These could be moved into into sanitizers. If so, then when the AMP plugin is running there would be no need to call `Document::fromHtml( $markup, get_bloginfo( 'charset' ) )` above. As it stands right now, the entire document is getting parsed and serialized twice when the AMP plugin is active.

Reply:
Due to the prefixing I would need to come up with a way to use the same code for both, as I need two classes that extend different base sanitizers. Maybe a trait?

Reply:
Good point. That should work. So then you'd have a single copy of the traits but then each would have two corresponding class "wrappers" that extend the AMP sanitizer and the Stories sanitizer respectively.

Reply:
Works well so far, but to completely get rid of the `Document::fromHtml` call I also need to move `add_poster_images` to the sanitizer which currently relies on `$this->story` in the `HTML` class to get the current story's poster images. Need to think of a good solution here.

Reply:
See https://github.com/google/web-stories-wp/pull/4513#pullrequestreview-502568064.